File: exercism-solutions/julia/luhn/luhn.jl

isdigit(x) = Int('0') <= Int(x) <= Int('9')

toint(x) = Int(x) - 48

function luhn(number)
    number = replace(number, " " => "")

    if length(number) < 2
        return false
    end

    sum = 0
    for (i, ch) in enumerate(reverse(number))
        x = Int(ch) - 48
        if x < 0 || x > 9
            return false
        end
        if i % 2 == 0
            x *= 2
            x = x > 9 ? x - 9 : x
        end
        sum += x
    end
    return rem(sum, 10) == 0
end

File: exercism-solutions/clojure/luhn/src/luhn.clj
(ns luhn)

(defn char->int [ch] (- (int ch) 48))

(defn luhn-transform [x]
  (#(if (> % 9) (- % 9) %) (* 2 x)))

(defn lunh-sum [digits]
  (reduce + (map-indexed
             #(if (even? %1)
                (char->int %2)
                (luhn-transform (char->int %2)))
             (reverse digits))))

(defn valid? [number]
  (let [digits (clojure.string/replace number #"\D" "")]
    (and
     (nil? (re-find #"[^\d ]" number))
     (> (count digits) 1)
     (= (mod (lunh-sum digits) 10) 0))))
